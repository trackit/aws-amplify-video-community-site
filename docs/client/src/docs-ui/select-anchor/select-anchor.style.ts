import {css} from "emotion";

export const selectAnchorStyle = css`
  position: relative;
  flex-grow: 2;

  display: block;
  user-select: none;

  stencil-route-link > a {
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: flex-start;
    padding: 0.5rem 0.75rem;
    border: 0.0625rem solid var(--border-color);
    color: var(--font-color);

    img {
      margin-right: 0.5rem;
      width: 1.75rem;
      height: 1.75rem;
    }

    &:hover {
      background-color: var(--bg-color-hover);
    }
  }
`;

export const currentlySelectedStyle = css`
  box-shadow: rgba(0, 0, 0, 0.09) 0.3125rem 0.3125rem 0 -0.0625rem;
  transition: all 0.6s cubic-bezier(0.165, 0.84, 0.44, 1);

  &:active {
    box-shadow: initial;
  }
`;

export const optionsStyle = css`
  visibility: hidden;
  position: absolute;
  background-color: var(--bg-color);
  box-shadow: rgba(0, 0, 0, 0.09) 0.3125rem 0.3125rem 0 -0.0625rem;
  width: 100%;
  z-index: 1;
  stencil-route-link > a {
    border: 0.0625rem solid var(--color-grey-md);
    border-top: 0;
  }
`;

export const showOptionsStyle = css`
  visibility: visible;
`;
