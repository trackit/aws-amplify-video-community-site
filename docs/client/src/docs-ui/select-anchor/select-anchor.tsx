import {
  Component,
  Host,
  h,
  Prop,
  State,
  Listen,
  Element,
  Watch,
  VNode,
} from "@stencil/core";
import {Page} from "../../api";
import {
  filterMetadataByOptionByName,
  FilterOptionMetadata,
} from "../../utils/filter-data";
import {getFilterKeyFromPage} from "../../utils/filters";
import {
  currentlySelectedStyle,
  optionsStyle,
  selectAnchorStyle,
  showOptionsStyle,
} from "./select-anchor.style";
import {SelectedFilters} from "../../docs-ui/page/page.types";
import {pageContext} from "../../docs-ui/page/page.context";
import {sidebarLayoutContext} from "../../amplify-ui/sidebar-layout/sidebar-layout.context";
import {ToggleInView} from "../../amplify-ui/sidebar-layout/sidebar-layout.types";
import {MQTablet} from "../../amplify-ui/styles/media";

@Component({tag: "docs-select-anchor"})
export class DocsSelectAnchor {
  @Element() element: HTMLElement;
  /*** the globally-selected filter state */
  @Prop() readonly selectedFilters?: SelectedFilters;
  /** the current page's data */
  @Prop() readonly page?: Page;
  /*** whether or not the menu is in view */
  @Prop() readonly inView?: boolean;
  /*** the menu toggle */
  @Prop() readonly toggleInView: ToggleInView;

  @State() showOptions = false;
  @State() sortedVersions?: [string, string][];

  selectedOption: SelectedFilters[keyof SelectedFilters];
  selectedOptionMetadata?: FilterOptionMetadata;
  options?: VNode[];

  toggleShowOptions = () => {
    this.showOptions = !this.showOptions;
  };

  toggleShowOptionsAndMenuInView = () => {
    this.toggleShowOptions();

    // Slice off the "@media " string at the start for use in JS instead of CSS
    const MQTabletJS = MQTablet.substring(6);
    const isMobile = !window.matchMedia(MQTabletJS).matches;
    if (isMobile && this.inView) {
      this.toggleInView();
    }
  };

  componentWillLoad() {
    this.computeVersions();
  }

  @Watch("page")
  computeAll() {
    this.computeOptionVNodes();
    this.computeVersions();
  }

  computeOptionVNodes() {
    const filterKey = this.page && getFilterKeyFromPage(this.page);
    this.selectedOption = filterKey && this.selectedFilters?.[filterKey];
    this.selectedOptionMetadata =
      filterKey &&
      this.selectedOption &&
      (filterMetadataByOptionByName[filterKey][
        this.selectedOption
      ] as FilterOptionMetadata);

    this.options =
      this.sortedVersions &&
      (this.sortedVersions.map(([filterValue, filterRoute]) => {
        if (filterValue !== this.selectedOption) {
          const optionMetadata =
            filterKey &&
            (filterMetadataByOptionByName[filterKey][
              filterValue
            ] as FilterOptionMetadata);
          return (
            optionMetadata && (
              <stencil-route-link
                key={filterValue}
                url={filterRoute}
                onClick={this.toggleShowOptionsAndMenuInView}
              >
                <img
                  src={optionMetadata.graphicURI}
                  alt={`${optionMetadata.label} Logo`}
                />
                <span>{optionMetadata.label}</span>
              </stencil-route-link>
            )
          );
        }
      }) as VNode[] | undefined);
  }

  computeVersions() {
    if (this.page?.versions) {
      const entries = Object.entries(this.page.versions);
      entries.sort(([a], [b]) => (a > b ? 1 : a < b ? -1 : 0));
      this.sortedVersions = entries;
      this.computeOptionVNodes();
    }
  }

  @Listen("click", {target: "window"})
  closeOnOuterClick(e: Event) {
    if (!(e && this.element.contains(e.target as HTMLElement))) {
      this.showOptions = false;
    }
  }

  render() {
    return (
      this.selectedOptionMetadata &&
      this.selectedOption && (
        <Host class={selectAnchorStyle}>
          <div class={currentlySelectedStyle}>
            <stencil-route-link
              activeClass={undefined}
              onClick={this.toggleShowOptions}
            >
              <img
                src={this.selectedOptionMetadata.graphicURI}
                alt={`${this.selectedOptionMetadata.label} Logo`}
              />
              <span>{this.selectedOptionMetadata.label}</span>
            </stencil-route-link>
          </div>
          <div
            class={{[optionsStyle]: true, [showOptionsStyle]: this.showOptions}}
          >
            {this.options}
          </div>
        </Host>
      )
    );
  }
}

pageContext.injectProps(DocsSelectAnchor, ["selectedFilters"]);
sidebarLayoutContext.injectProps(DocsSelectAnchor, ["inView", "toggleInView"]);
