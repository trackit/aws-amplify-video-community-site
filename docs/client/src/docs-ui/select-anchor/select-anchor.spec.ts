import {DocsSelectAnchor} from "./select-anchor";
import {newSpecPage} from "@stencil/core/testing";

describe("docs-select-anchor", () => {
  describe("Component logic", () => {
    let selectAnchor: DocsSelectAnchor;
    beforeEach(() => (selectAnchor = new DocsSelectAnchor()));

    it("should init `page` as `undefined`", () =>
      expect(selectAnchor.page).toBeUndefined());

    it("should init `showOptions` as `false`", () =>
      expect(selectAnchor.showOptions).toBe(false));
  });

  describe("Render logic", () => {
    it("should render", async () => {
      expect(
        (
          await newSpecPage({
            components: [DocsSelectAnchor],
            html: `<docs-select-anchor />`,
          })
        ).root,
      ).toMatchSnapshot();
    });
  });
});
