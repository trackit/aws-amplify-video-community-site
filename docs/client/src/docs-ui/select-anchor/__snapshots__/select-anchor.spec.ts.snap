// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`docs-select-anchor Render logic should render 1`] = `<docs-select-anchor></docs-select-anchor>`;
